agent/agents/postgres/parser/parser.go

```diff
@@ -45,122 +44,36 @@ func ExtractTables(query string) (tables []string, err error) {
-		if !tableAdded && !tableExcluded {
-			tables = append(tables, tableName)
-			tableNames[tableName] = true
-		}
+	for _, v := range extract(jsonTree, `"ctename":"`, `"`) {
+		delete(tableNames, v)
 	}
 
+	tables = []string{}
+	for k := range tableNames {
+		tables = append(tables, k)
+	}
 	sort.Strings(tables)
 
 	return
 }
 
-func extractTableNames(stmts ...json.RawMessage) ([]string, []string) {
-	var tables, excludeTables []string
-	for _, input := range stmts {
-		if input == nil || string(input) == "null" || !(strings.HasPrefix(string(input), "{") || strings.HasPrefix(string(input), "[")) {
-			continue
-		}
-
-		if strings.HasPrefix(string(input), "[") {
-			var list []json.RawMessage
-			if err := json.Unmarshal(input, &list); err != nil {
-				panic(err)
-			}
-			foundTables, tmpExcludeTables := extractTableNames(list...)
-			tables = append(tables, foundTables...)
-			excludeTables = append(excludeTables, tmpExcludeTables...)
-			continue
-		}
-
-		var nodeMap map[string]json.RawMessage
-		if err := json.Unmarshal(input, &nodeMap); err != nil {
-			panic(err)
-		}
-
-		for nodeType, jsonText := range nodeMap {
-			if jsonText == nil || string(jsonText) == "null" {
-				continue
-			}
-
-			var foundTables, tmpExcludeTables []string
-			switch nodeType {
-			case "RangeVar":
-				var outNode pgquerynodes.RangeVar
-				if err := json.Unmarshal(jsonText, &outNode); err != nil {
-					panic(err)
-				}
-				tables = append(tables, *outNode.Relname)
-				continue
-
-			case "List":
-				foundTables, tmpExcludeTables = extractTableNames(jsonText)
-
-			default:
-				var nm map[string]json.RawMessage
-				if err := json.Unmarshal(jsonText, &nm); err != nil {
-					panic(err)
-				}
-
-				switch nodeType {
-				case "RangeVar":
-				case "CommonTableExpr":
-					foundTables, tmpExcludeTables = extractTableNames(nm["ctequery"])
-					cteName := string(nm["ctename"])
-					cteName = strings.TrimPrefix(cteName, `"`)
-					cteName = strings.TrimSuffix(cteName, `"`)
-					tmpExcludeTables = append(tmpExcludeTables, cteName)
-
-				case "RawStmt":
-					foundTables, tmpExcludeTables = extractTableNames(nm["stmt"])
-				case "SelectStmt":
-					foundTables, tmpExcludeTables = extractTableNames(nm["fromClause"], nm["whereClause"], nm["withClause"], nm["larg"], nm["rarg"])
-				case "InsertStmt":
-					foundTables, tmpExcludeTables = extractTableNames(nm["relation"], nm["selectStmt"], nm["withClause"])
-				case "UpdateStmt":
-					foundTables, tmpExcludeTables = extractTableNames(nm["relation"], nm["fromClause"], nm["whereClause"], nm["withClause"])
-				case "DeleteStmt":
-					foundTables, tmpExcludeTables = extractTableNames(nm["relation"], nm["whereClause"], nm["withClause"])
-
-				case "JoinExpr":
-					foundTables, tmpExcludeTables = extractTableNames(nm["larg"], nm["rarg"])
+func extract(query, pre, post string) []string {
+	re := regexp.MustCompile(fmt.Sprintf("(%s)(.*?)(%s)", pre, post))
```

Comment:
Better to use `regexp.Compile` here and handle error. Otherwise it's easy to get panic here because of dynamic expression.